### PR TITLE
Add inlines for overflow detection

### PR DESCRIPTION
**Pull request by smcv**
## Description

It's common to do arithmetic to work out how much space to allocate for something. This can be a problem if the inputs to the arithmetic are attacker-supplied and could overflow (as in libsdl-org/SDL_ttf#187 if using a malicious attacker-supplied font). Signed integer overflow is undefined behaviour, which could make anything happen, including a crash or a security vulnerability. Unsigned overflow is defined to wrap around, but that's exactly what we don't want when dealing with memory allocations, because it would be Very Bad to allocate less memory than we thought we had.

We can defend against this sort of thing by doing explicit bounds checks, but because signed overflow is UB, it's necessary to avoid actually doing the computation that would overflow. Recent versions of gcc (and I think clang?) have some builtins that are useful for this, and it's reasonably straightforward to emulate them for other compilers by shuffling around the expression to ensure no single step overflows.

## Existing Issue(s)

No specific issue; defending against issues like libsdl-org/SDL_ttf#187

## Commits

* stdinc: Add overflow-checking add and multiply for size_t
    
    This can be used to check whether untrusted sizes would cause overflow
    when used to calculate how much memory is needed.

* test: Add a unit test for overflow detection

* cpuinfo: Set padding to 0 if none is needed
    
    It'll be simpler to use overflow detection after this refactor.

* cpuinfo: Check for overflow in SIMD allocation
    
    If the size to be allocated is very large and untrusted, then adding
    the padding etc. might be enough to cause unsigned overflow, after
    which a very small amount of memory will be allocated.

### CODE REVIEW

File: include/SDL_stdinc.h [hunk 1]
```diff
@@ -728,6 +734,60 @@ SDL_FORCE_INLINE void *SDL_memcpy4(SDL_OUT_BYTECAP(dwords*4) void *dst, SDL_IN_B
...
+{
+    if (a != 0 && b > SDL_SIZE_MAX / a) {
+        return -1;
+    }
+    *ret = a * b;
+    return 0;
+}
+
+#if _SDL_HAS_BUILTIN(__builtin_mul_overflow)
+SDL_FORCE_INLINE int _SDL_size_mul_overflow_builtin (size_t a,
```
Comment (smcv) [inline include/SDL_stdinc.h:755]: I should add a comment in the code, but for now, a comment on the PR instead:

The reason the builtin is wrapped in an inline is that `__builtin_mul_overflow()` is generic across types (!) but I'm fairly sure we want this to be consistently interpreting its arguments as `size_t` for consistency with the non-builtin.

Comment (smcv) [reply include/SDL_stdinc.h:755]: Follow-up for this is on #5644
